### PR TITLE
Generate seed for each machine in each fold

### PR DESCRIPTION
**Pull request by gf712**
@karlnapf is this what you had in mind for xvalidation?

### CODE REVIEW

File: src/shogun/evaluation/CrossValidation.cpp [hunk 1]
```diff
@@ -95,6 +96,12 @@ float64_t CrossValidation::evaluate_one_run(int64_t index) const
 
 	SGVector<float64_t> results(num_subsets);
 
+	std::vector<int32_t> seed_vector(num_subsets);
+	UniformIntDistribution<int32_t> seed_dist(std::numeric_limits<int32_t>::min(), 
+		std::numeric_limits<int32_t>::max());
```
Comment (gf712) [inline src/shogun/evaluation/CrossValidation.cpp]: @vigsterkr not sure how else to generate int32_t values. Should seed be [0, 2^31-1] or [-2^31, 2^31-1]?

Comment (vigsterkr) [reply src/shogun/evaluation/CrossValidation.cpp]: why dont you just use the prng directly?

Comment (gf712) [reply src/shogun/evaluation/CrossValidation.cpp]: because prng returns a uint64_t, which then would get truncated when casted to int32 no?

Comment (vigsterkr) [reply src/shogun/evaluation/CrossValidation.cpp]: why do you need signed int32_t? i mean it's gonna be a seed for another prng no?
for mersenne prng the seed should be actually unsigned 💃 so we have a bug in the Seedable :P

File: src/shogun/evaluation/CrossValidation.cpp [hunk 2]
```diff
@@ -95,6 +96,12 @@ float64_t CrossValidation::evaluate_one_run(int64_t index) const
 
 	SGVector<float64_t> results(num_subsets);
 
+	std::vector<int32_t> seed_vector(num_subsets);
```
Comment (vigsterkr) [inline src/shogun/evaluation/CrossValidation.cpp]: `std::vector<decltype(m_prng::result_type)>`?

File: src/shogun/evaluation/CrossValidation.cpp (hunk 1 above)
Comment (gf712) [reply src/shogun/evaluation/CrossValidation.cpp]: So Seedable needs to know what the input type of the prng is?

**Comment by karlnapf** [comment]
> @karlnapf is this what you had in mind for xvalidation?

yes! :)  I guess only some details need updating.

File: src/shogun/mathematics/RandomNamespace.h [hunk 3]
```diff
@@ -105,6 +105,25 @@ namespace shogun
...
+		 * @param num_elements number of elements to generate
+		 */
+		template <typename PRNG, typename T>
+		std::vector<typename std::remove_reference_t<PRNG>::result_type> 
+		generate_array(PRNG&& prng, const T& num_elements) 
+		{
+			std::vector<typename std::remove_reference_t<PRNG>::result_type> result;
+			result.reserve(num_elements);
+			std::generate_n(std::back_inserter(result), num_elements, 
+				[prng = std::forward<PRNG>(prng)]() mutable {
```
Comment (vigsterkr) [inline src/shogun/mathematics/RandomNamespace.h:121]: badass!

File: src/shogun/mathematics/Seedable.h [hunk 4]
```diff
@@ -62,7 +62,7 @@ namespace shogun
 			random::seed(object, m_seed);
 		}
 
-		int32_t m_seed;
+		int64_t m_seed;
```
Comment (vigsterkr) [inline src/shogun/mathematics/Seedable.h:65]: these lines should be changed as well to `int64_t`:
https://github.com/shogun-toolbox/shogun/blob/167880c1ea5bc62ec4906f48a86d72427eb9a014/src/shogun/mathematics/Seedable.h#L24
https://github.com/shogun-toolbox/shogun/blob/167880c1ea5bc62ec4906f48a86d72427eb9a014/src/shogun/mathematics/Seedable.h#L31
https://github.com/shogun-toolbox/shogun/blob/167880c1ea5bc62ec4906f48a86d72427eb9a014/src/shogun/mathematics/Seedable.h#L68

**Review by vigsterkr — CHANGES_REQUESTED** [review]

File: src/shogun/base/SGObject.h [hunk 5]
```diff
@@ -844,6 +856,13 @@ class SGObject: public std::enable_shared_from_this<SGObject>
 					visitor->m_value = *value;	
 				}
 			);
+			if constexpr(std::is_arithmetic_v<T> && !std::is_same_v<T, bool>) 
+			{
+				Any::register_visitor<T, ParameterGetterInterface<int32_t, int32_t>>(
```
Comment (karlnapf) [inline src/shogun/base/SGObject.h:861]: that is nice to do that down here!

Comment (gf712) [reply src/shogun/base/SGObject.h:861]: maybe.. But this is getting a bit out of hand :D I thought about separating all of this out and adding to a mixin, but I think the resulting multiple inheritance might cause headaches..

**Pull request closed without merging**